### ISSUE TITLE
Redesign adjoint / transpose interface and algorithms

### ISSUE THREAD

**Issue by fritzo** [pyro-ppl/funsor#446]
## Reading list

- https://popl21.sigplan.org/details/lafi-2021-papers/9/Decomposing-reverse-mode-automatic-differentiation
- [The Simple Essence of Automatic Differentiation](https://arxiv.org/pdf/1804.00746.pdf)

**Comment by ordabayevy** [comment]
Just throwing out an idea here. Another perspective on two different semi-rings (ops.add, ops.mul) and (ops.logaddexp, ops.add) is to view the latter as operations lifted by a `Log` Functor. In Haskell notation:

```hs
instance Functor Log where
    fmap f :: Log a -> Log b
    fmap f (Log a) = Log (f a)

...

fmap (*) (Log (x, y)) = Log (x*y) = Log x + Log y
fmap (+)  = logaddexp
```

These two different types could be represented by something like `Factor` and `Logfactor` funsors which can be used for pattern matching.


As a side note, can Funsor be used to represent partially applied functions?

**Comment by fritzo** [comment]
@ordabayevy yes I think your `Log` functor relates the semiring pair `((add,mul), (logaddexp,add))` and the pair `((max,mul),(max,add))`. However no functor relates e.g. the pair `((add,mul), (max,add))`.

Indeed `Funsor`s can represent functions, and funsor substitution represents partially applied functions.
